bumble/profiles/hap.py

```diff
@@ -0,0 +1,599 @@
+                is_last,
+            ]
+        ) + additional_parameters_bytes
+
+
+class PresetChangedOperationDeleted(PresetChangedOperation):
+    def __init__(self, index):
+        self.change_id = PresetChangedOperation.ChangeId.PRESET_RECORD_DELETED
+        self.additional_parameters = index
+
+
+class PresetChangedOperationAvailable(PresetChangedOperation):
+    def __init__(self, index):
+        self.change_id = PresetChangedOperation.ChangeId.PRESET_RECORD_AVAILABLE
+        self.additional_parameters = index
+
+
+class PresetChangedOperationUnavailable(PresetChangedOperation):
+    def __init__(self, index):
+        self.change_id = PresetChangedOperation.ChangeId.PRESET_RECORD_UNAVAILABLE
+        self.additional_parameters = index
+
+
+@dataclass
+class PresetRecord:
+    '''See Hearing Access Service 2.8. Preset record.'''
+
+    @dataclass
+    class Property:
+        class Writable(OpenIntEnum):
+            CANNOT_BE_WRITTEN = 0b0
+            CAN_BE_WRITTEN = 0b1
+
+        class IsAvailable(OpenIntEnum):
+            IS_UNAVAILABLE = 0b0
+            IS_AVAILABLE = 0b1
+
+        writable: Writable = Writable.CAN_BE_WRITTEN
+        is_available: IsAvailable = IsAvailable.IS_AVAILABLE
+
+        def __bytes__(self) -> bytes:
+            return bytes([self.writable | (self.is_available << 1)])
+
+    index: int
+    name: str
+    properties: Property = field(default_factory=Property)
+
+    def __bytes__(self) -> bytes:
+        return bytes([self.index]) + bytes(self.properties) + self.name.encode('utf-8')
+
+    def is_available(self) -> bool:
+        return (
+            self.properties.is_available
+            == PresetRecord.Property.IsAvailable.IS_AVAILABLE
+        )
+
+
+# -----------------------------------------------------------------------------
+# Server
+# -----------------------------------------------------------------------------
+class HearingAccessService(gatt.TemplateService):
+    UUID = gatt.GATT_HEARING_ACCESS_SERVICE
+
+    hearing_aid_features_characteristic: gatt.Characteristic
+    hearing_aid_preset_control_point: gatt.Characteristic
+    active_preset_index_characteristic: gatt.Characteristic
+    active_preset_index: int = 0x00
+    # TODO validate address type → public_address ?
+    active_preset_index_per_device: Dict[Address, int] = {}
+
+    device: Device
+
+    server_features: HearingAidFeatures
+    preset_records: Dict[int, PresetRecord] = {}  # key is the preset index
+    read_presets_request_in_progress: bool = False
+
+    preset_changed_operations_history_per_device: Dict[
+        Address, List[PresetChangedOperation]
+    ] = {}
+
+    # Keep an updated list of connected client to send notification to
+    currently_connected_clients: Set[Connection] = set()
+
+    def __init__(
+        self, device: Device, features: HearingAidFeatures, presets: List[PresetRecord]
+    ) -> None:
+        # 2.5. GATT sub-procedure requirements
+        assert att.ATT_DEFAULT_MTU >= 49
```

Comment:
Though Python doesn't forbid overwrite constants, this does violate the convention. Is it possible to check the MTU in use from `device`, instead of checking and chaning the constant?

Reply:
I am not sure how.
I intended this check to be a "friendly warning" to the developer that forgot to negotiate it.
Because it will works without setting the MTU, but some packet will be silently truncated.

If we check the device mtu, we need to have an already negotiated MTU when creating the service, which is not the case.
The service is created, then added to the advertisement of the device then the connection is established and then the MTU is negotiated.

I was hoping that by setting the ATT_DEFAULT_MTU to another value I wouldn't need a manual negotiation but it appear that the MTU negotiation is still needed.

I could simply remove the assert and let other find out as I don't expect this issue to be HAS specific

Reply:
I will suggest to do this on the caller side, and inside HAS we only check and pop warnings?